sys/kern/tty.c

```diff
@@ -563,21 +609,47 @@ static int tty_write(vnode_t *v, uio_t *uio, int ioflags) {
 
   uio->uio_offset = 0; /* This device does not support offsets. */
 
-  uint8_t c;
   WITH_MTX_LOCK (&tty->t_lock) {
-    while (uio->uio_resid > 0) {
-      error = uiomove(&c, 1, uio);
-      if (error)
-        break;
-      tty->t_rocount = 0;
-      tty_output(tty, c);
-    }
-    tty_notify_out(tty);
+    /* Wait for our turn. */
+    while (tty->t_flags & TF_OUT_BUSY)
+      cv_wait(&tty->t_serialize_cv, &tty->t_lock);
+    tty->t_flags |= TF_OUT_BUSY;
+
+    error = tty_do_write(tty, uio);
```

Comment:
Perhaps a comment would be helpful here. It's not obvious that `tty_do_write` may release `t_lock` and setting `TF_OUT_BUSY` is well justified.

Reply:
Done.